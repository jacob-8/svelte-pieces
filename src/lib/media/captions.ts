import { parse, type X2jOptions } from 'fast-xml-parser';
import he from 'he';

export interface YoutubeCaptionTrack {
  id: number;
  name: string;
  langCode: string;
  langOriginal: string;
  langTranslated: string;
  langDefault: boolean;
}

export interface YoutubeCaption {
  text: string;
  start: number;
  duration?: number;
}

export const getTracks = async (videoId: string): Promise<YoutubeCaptionTrack[]> => {
  const response = await fetch(
    `https://cors-proxy.peoples.workers.dev/?https://python.zerotohero.ca/timedtext?v=${videoId}&type=list`
  );
  const xml = await response.text();
  return parseTracksXml(xml);
};

export const getCaptions = async (
  videoId: string,
  { langCode, name }: YoutubeCaptionTrack
): Promise<YoutubeCaption[]> => {
  const url = `https://cors-proxy.peoples.workers.dev/?https://python.zerotohero.ca/timedtext?v=${videoId}&lang=${langCode}${
    name ? '&name=' + name : ''
  }&fmt=srv3`; //https://python.zerotohero.ca/timedtext?v=7XeYmKhrK9Q&lang=zh-Hans&fmt=srv3
  // if (name) url += `&name=${name}`;
  const response = await fetch(url);
  const xml = await response.text();
  return parseSrv3CaptionsXml(xml);
};

const options: Partial<X2jOptions> = {
  ignoreAttributes: false,
  ignoreNameSpace: true, // ?
  allowBooleanAttributes: true,
  parseNodeValue: true,
  textNodeName: 'text',
  parseAttributeValue: true,
  attributeNamePrefix: '',
  trimValues: true,
  parseTrueNumberOnly: false,
  arrayMode: false, //"strict"
  attrValueProcessor: (val, attrName) => he.decode(val, { isAttributeValue: true }), //default is a=>a
  tagValueProcessor: (val, tagName) => he.decode(val), //default is a=>a
};

export const parseTracksXml = (xml: string): YoutubeCaptionTrack[] => {
  const res = parse(xml, options);
  if (!res.transcript_list) {
    return [];
  }
  let tracks = res.transcript_list.track;
  if (!Array.isArray(res.transcript_list.track)) {
    tracks = [tracks];
  }

  return tracks.map((attrs) => {
    return {
      id: Number(attrs.id),
      name: attrs.name,
      langCode: attrs.lang_code,
      langOriginal: attrs.lang_original,
      langTranslated: attrs.lang_translated,
      langDefault: attrs.lang_default === true,
    };
  });
};

// deprecated with the passing of YouTube's API
// export const parseCaptionsXml = (xml: string): YoutubeCaption[] => {
//   const res = parse(xml, options);

//   if (!res.transcript || !Array.isArray(res.transcript.text)) {
//     return [];
//   }

//   return res.transcript.text
//     .map((item) => {
//       return {
//         text: item.text || '',
//         start: Number(item.start),
//         duration: item.dur ? Number(item.dur) : undefined,
//       };
//     })
//     .filter((item) => item.text);
// };

export const parseSrv3CaptionsXml = (xml: string): YoutubeCaption[] => {
  const res = parse(xml, options);

  if (!(res?.timedtext?.body?.p && Array.isArray(res.timedtext.body.p))) {
    return [];
  }

  const lines = res.timedtext.body.p;
  return lines
    .map((item) => {
      return {
        text: item.text || '',
        start: Number(item.t) / 1000,
        duration: item.d ? Number(item.d) / 1000 : undefined,
      };
    })
    .filter((item) => item.text);
  // see https://github.com/longjiang/zerotohero-nuxt/blob/master/lib/youtube.js#L40-L54 if having trouble with empty lines or out of order
};

// from @os-team/youtube-captions though its sax depenency is only for Node so am using fast-xml-parser instead to cover browser usage
// useful links: https://stackoverflow.com/questions/23665343/get-closed-caption-cc-for-youtube-video/30601863

export const findCurrentCaption = (captions: YoutubeCaption[], time: number) => {
  return captions.find(({ start, duration }) => start <= time && start + duration > time);
};
