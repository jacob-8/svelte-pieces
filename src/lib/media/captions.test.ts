import { parseTracksXml } from './captions';
// import fetch from 'isomorphic-unfetch';
// improve fetch mock with https://kentcdodds.com/blog/stop-mocking-fetch

// test('getTracks should get tracks', async () => {
//   global.fetch = fetch;
//   expect(await getTracks('rokGy0huYEA')).toMatchInlineSnapshot();
// });

test('parseTracksXml returns array even when there is 1 track', () => {
  expect(
    parseTracksXml(
      '<?xml version="1.0" encoding="utf-8" ?><transcript_list docid="14931616755950075552"><track id="0" name="" lang_code="en" lang_original="English" lang_translated="English" lang_default="true"/></transcript_list>'
    )
  ).toHaveLength(1);
});

test('parseTracksXml returns array of tracks', () => {
  expect(
    parseTracksXml(`<?xml version="1.0" encoding="utf-8" ?><transcript_list docid="12576590904572141632"><track id="0" name="" lang_code="ar" lang_original="العربية" lang_translated="Arabic"/><track id="2" name="" lang_code="bn" lang_original="বাংলা" lang_translated="Bangla"/><track id="1" name="" lang_code="bg" lang_original="Български" lang_translated="Bulgarian"/><track id="3" name="" lang_code="ca" lang_original="Català" lang_translated="Catalan"/><track id="54" name="" lang_code="zh-HK" lang_original="中文（香港）" lang_translated="Chinese (Hong Kong)"/><track id="53" name="" lang_code="zh-CN" lang_original="中文（简体）" lang_translated="Chinese (Simplified)"/><track id="55" name="" lang_code="zh-TW" lang_original="中文（繁體）" lang_translated="Chinese (Traditional)"/><track id="21" name="" lang_code="hr" lang_original="Hrvatski" lang_translated="Croatian"/><track 
    id="4" name="" lang_code="cs" lang_original="Čeština" lang_translated="Czech"/><track id="5" name="" lang_code="da" lang_original="Dansk" lang_translated="Danish"/><track id="33" name="" lang_code="nl-BE" lang_original="Nederlands (België)" lang_translated="Dutch (Belgium)"/><track id="34" name="" lang_code="nl-NL" lang_original="Nederlands (Nederland)" lang_translated="Dutch (Netherlands)"/><track id="8" name="" lang_code="en-GB" lang_original="English (United Kingdom)" lang_translated="English (United Kingdom)" lang_default="true"/><track id="9" name="CC" lang_code="en-US" lang_original="English (United States)" lang_translated="English (United States)"/><track id="12" name="" lang_code="et" lang_original="Eesti" lang_translated="Estonian"/><track id="15" name="" lang_code="fil" lang_original="Filipino" lang_translated="Filipino"/><track id="14" name="" 
    lang_code="fi" lang_original="Suomi" lang_translated="Finnish"/><track id="16" name="" lang_code="fr-BE" lang_original="Français (Belgique)" lang_translated="French (Belgium)"/><track id="17" name="" lang_code="fr-CA" lang_original="Français (Canada)" lang_translated="French (Canada)"/><track id="18" name="" lang_code="fr-FR" lang_original="Français (France)" lang_translated="French (France)"/><track id="6" name="" lang_code="de" 
    lang_original="Deutsch" lang_translated="German"/><track id="7" name="" lang_code="el" lang_original="Ελληνικά" lang_translated="Greek"/><track id="19" name="" lang_code="gu" lang_original="ગુજરાતી" lang_translated="G
    ujarati"/><track id="24" name="" lang_code="iw" lang_original="עברית" lang_translated="Hebrew"/><track id="20" name="" lang_code="hi" lang_original="हिन्दी" lang_translated="Hindi"/><track id="22" name="" lang_code="i
    d" lang_original="Indonesia" lang_translated="Indonesian"/><track id="23" name="" lang_code="it" lang_original="Italiano" lang_translated="Italian"/><track id="25" name="" lang_code="ja" lang_original="日本語" lang_translated="Japanese"/><track id="26" name="" lang_code="kn" lang_original="ಕನ್ನಡ" lang_translated="Kannada"/><track id="27" name="" lang_code="ko" lang_original="한국어" lang_translated="Korean"/><track id="29" name=""
     lang_code="lv" lang_original="Latviešu" lang_translated="Latvian"/><track id="28" name="" lang_code="lt" lang_original="Lietuvių" lang_translated="Lithuanian"/><track id="32" name="" lang_code="ms" lang_original="Melayu" lang_translated="Malay"/><track id="30" name="" lang_code="ml" lang_original="മലയാളം" lang_translated="Malayalam"/><track id="31" name="" lang_code="mr" lang_original="मराठी" lang_translated="Marathi"/><track id="35" name="" lang_code="no" lang_original="Norsk" lang_translated="Norwegian"/><track id="13" name="" lang_code="fa" lang_original="فارسی" lang_translated="Persian"/><track id="37" name="" lang_code="pl" lang_original="Polski" lang_translated="Polish"/><track id="38" name="" lang_code="pt-BR" lang_original="Português (Brasil)" lang_translated="Portuguese (Brazil)"/><track id="39" name="" lang_code="pt-PT" lang_original="Português 
    (Portugal)" lang_translated="Portuguese (Portugal)"/><track id="36" name="" lang_code="pa" lang_original="ਪੰਜਾਬੀ" lang_translated="Punjabi"/><track id="40" name="" lang_code="ro" lang_original="Română" lang_translated
    ="Romanian"/><track id="41" name="" lang_code="ru" lang_original="Русский" lang_translated="Russian"/><track id="44" name="" lang_code="sr" lang_original="Српски" lang_translated="Serbian"/><track id="42" name="" lang_code="sk" lang_original="Slovenčina" lang_translated="Slovak"/><track id="43" name="" lang_code="sl" lang_original="Slovenščina" lang_translated="Slovenian"/><track id="10" name="" lang_code="es-419" lang_original="Español (Latinoamérica)" lang_translated="Spanish (Latin America)"/><track id="11" name="" lang_code="es-ES" lang_original="Español (España)" lang_translated="Spanish (Spain)"/><track id="46" name="" lang_code="sw" lang_original="Kiswahili" lang_translated="Swahili"/><track id="45" name="" lang_code="sv" lang_original="Svenska" lang_translated="Swedish"/><track id="47" name="" lang_code="ta" lang_original="தமிழ்" lang_translated="T
    amil"/><track id="48" name="" lang_code="te" lang_original="తెలుగు" lang_translated="Telugu"/><track id="49" name="" lang_code="th" lang_original="ไทย" lang_translated="Thai"/><track id="50" name="" lang_code="tr" lan
    g_original="Türkçe" lang_translated="Turkish"/><track id="51" name="" lang_code="uk" lang_original="Українська" lang_translated="Ukrainian"/><track id="52" name="" lang_code="vi" lang_original="Tiếng Việt" lang_translated="Vietnamese"/></transcript_list>`)
  ).toMatchSnapshot();
});
