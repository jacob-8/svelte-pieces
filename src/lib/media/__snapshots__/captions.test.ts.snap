// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseTracksXml returns array of tracks 1`] = `
[
  {
    "id": 0,
    "langCode": "ar",
    "langDefault": false,
    "langOriginal": "العربية",
    "langTranslated": "Arabic",
    "name": "",
  },
  {
    "id": 2,
    "langCode": "bn",
    "langDefault": false,
    "langOriginal": "বাংলা",
    "langTranslated": "Bangla",
    "name": "",
  },
  {
    "id": 1,
    "langCode": "bg",
    "langDefault": false,
    "langOriginal": "Български",
    "langTranslated": "Bulgarian",
    "name": "",
  },
  {
    "id": 3,
    "langCode": "ca",
    "langDefault": false,
    "langOriginal": "Català",
    "langTranslated": "Catalan",
    "name": "",
  },
  {
    "id": 54,
    "langCode": "zh-HK",
    "langDefault": false,
    "langOriginal": "中文（香港）",
    "langTranslated": "Chinese (Hong Kong)",
    "name": "",
  },
  {
    "id": 53,
    "langCode": "zh-CN",
    "langDefault": false,
    "langOriginal": "中文（简体）",
    "langTranslated": "Chinese (Simplified)",
    "name": "",
  },
  {
    "id": 55,
    "langCode": "zh-TW",
    "langDefault": false,
    "langOriginal": "中文（繁體）",
    "langTranslated": "Chinese (Traditional)",
    "name": "",
  },
  {
    "id": 21,
    "langCode": "hr",
    "langDefault": false,
    "langOriginal": "Hrvatski",
    "langTranslated": "Croatian",
    "name": "",
  },
  {
    "id": 4,
    "langCode": "cs",
    "langDefault": false,
    "langOriginal": "Čeština",
    "langTranslated": "Czech",
    "name": "",
  },
  {
    "id": 5,
    "langCode": "da",
    "langDefault": false,
    "langOriginal": "Dansk",
    "langTranslated": "Danish",
    "name": "",
  },
  {
    "id": 33,
    "langCode": "nl-BE",
    "langDefault": false,
    "langOriginal": "Nederlands (België)",
    "langTranslated": "Dutch (Belgium)",
    "name": "",
  },
  {
    "id": 34,
    "langCode": "nl-NL",
    "langDefault": false,
    "langOriginal": "Nederlands (Nederland)",
    "langTranslated": "Dutch (Netherlands)",
    "name": "",
  },
  {
    "id": 8,
    "langCode": "en-GB",
    "langDefault": true,
    "langOriginal": "English (United Kingdom)",
    "langTranslated": "English (United Kingdom)",
    "name": "",
  },
  {
    "id": 9,
    "langCode": "en-US",
    "langDefault": false,
    "langOriginal": "English (United States)",
    "langTranslated": "English (United States)",
    "name": "CC",
  },
  {
    "id": 12,
    "langCode": "et",
    "langDefault": false,
    "langOriginal": "Eesti",
    "langTranslated": "Estonian",
    "name": "",
  },
  {
    "id": 15,
    "langCode": "fil",
    "langDefault": false,
    "langOriginal": "Filipino",
    "langTranslated": "Filipino",
    "name": "",
  },
  {
    "id": 14,
    "langCode": "fi",
    "langDefault": false,
    "langOriginal": "Suomi",
    "langTranslated": "Finnish",
    "name": "",
  },
  {
    "id": 16,
    "langCode": "fr-BE",
    "langDefault": false,
    "langOriginal": "Français (Belgique)",
    "langTranslated": "French (Belgium)",
    "name": "",
  },
  {
    "id": 17,
    "langCode": "fr-CA",
    "langDefault": false,
    "langOriginal": "Français (Canada)",
    "langTranslated": "French (Canada)",
    "name": "",
  },
  {
    "id": 18,
    "langCode": "fr-FR",
    "langDefault": false,
    "langOriginal": "Français (France)",
    "langTranslated": "French (France)",
    "name": "",
  },
  {
    "id": 6,
    "langCode": "de",
    "langDefault": false,
    "langOriginal": "Deutsch",
    "langTranslated": "German",
    "name": "",
  },
  {
    "id": 7,
    "langCode": "el",
    "langDefault": false,
    "langOriginal": "Ελληνικά",
    "langTranslated": "Greek",
    "name": "",
  },
  {
    "id": 19,
    "langCode": "gu",
    "langDefault": false,
    "langOriginal": "ગુજરાતી",
    "langTranslated": "G     ujarati",
    "name": "",
  },
  {
    "id": 24,
    "langCode": "iw",
    "langDefault": false,
    "langOriginal": "עברית",
    "langTranslated": "Hebrew",
    "name": "",
  },
  {
    "id": 20,
    "langCode": "hi",
    "langDefault": false,
    "langOriginal": "हिन्दी",
    "langTranslated": "Hindi",
    "name": "",
  },
  {
    "id": 22,
    "langCode": "i     d",
    "langDefault": false,
    "langOriginal": "Indonesia",
    "langTranslated": "Indonesian",
    "name": "",
  },
  {
    "id": 23,
    "langCode": "it",
    "langDefault": false,
    "langOriginal": "Italiano",
    "langTranslated": "Italian",
    "name": "",
  },
  {
    "id": 25,
    "langCode": "ja",
    "langDefault": false,
    "langOriginal": "日本語",
    "langTranslated": "Japanese",
    "name": "",
  },
  {
    "id": 26,
    "langCode": "kn",
    "langDefault": false,
    "langOriginal": "ಕನ್ನಡ",
    "langTranslated": "Kannada",
    "name": "",
  },
  {
    "id": 27,
    "langCode": "ko",
    "langDefault": false,
    "langOriginal": "한국어",
    "langTranslated": "Korean",
    "name": "",
  },
  {
    "id": 29,
    "langCode": "lv",
    "langDefault": false,
    "langOriginal": "Latviešu",
    "langTranslated": "Latvian",
    "name": "",
  },
  {
    "id": 28,
    "langCode": "lt",
    "langDefault": false,
    "langOriginal": "Lietuvių",
    "langTranslated": "Lithuanian",
    "name": "",
  },
  {
    "id": 32,
    "langCode": "ms",
    "langDefault": false,
    "langOriginal": "Melayu",
    "langTranslated": "Malay",
    "name": "",
  },
  {
    "id": 30,
    "langCode": "ml",
    "langDefault": false,
    "langOriginal": "മലയാളം",
    "langTranslated": "Malayalam",
    "name": "",
  },
  {
    "id": 31,
    "langCode": "mr",
    "langDefault": false,
    "langOriginal": "मराठी",
    "langTranslated": "Marathi",
    "name": "",
  },
  {
    "id": 35,
    "langCode": "no",
    "langDefault": false,
    "langOriginal": "Norsk",
    "langTranslated": "Norwegian",
    "name": "",
  },
  {
    "id": 13,
    "langCode": "fa",
    "langDefault": false,
    "langOriginal": "فارسی",
    "langTranslated": "Persian",
    "name": "",
  },
  {
    "id": 37,
    "langCode": "pl",
    "langDefault": false,
    "langOriginal": "Polski",
    "langTranslated": "Polish",
    "name": "",
  },
  {
    "id": 38,
    "langCode": "pt-BR",
    "langDefault": false,
    "langOriginal": "Português (Brasil)",
    "langTranslated": "Portuguese (Brazil)",
    "name": "",
  },
  {
    "id": 39,
    "langCode": "pt-PT",
    "langDefault": false,
    "langOriginal": "Português      (Portugal)",
    "langTranslated": "Portuguese (Portugal)",
    "name": "",
  },
  {
    "id": 36,
    "langCode": "pa",
    "langDefault": false,
    "langOriginal": "ਪੰਜਾਬੀ",
    "langTranslated": "Punjabi",
    "name": "",
  },
  {
    "id": 40,
    "langCode": "ro",
    "langDefault": false,
    "langOriginal": "Română",
    "langTranslated": "Romanian",
    "name": "",
  },
  {
    "id": 41,
    "langCode": "ru",
    "langDefault": false,
    "langOriginal": "Русский",
    "langTranslated": "Russian",
    "name": "",
  },
  {
    "id": 44,
    "langCode": "sr",
    "langDefault": false,
    "langOriginal": "Српски",
    "langTranslated": "Serbian",
    "name": "",
  },
  {
    "id": 42,
    "langCode": "sk",
    "langDefault": false,
    "langOriginal": "Slovenčina",
    "langTranslated": "Slovak",
    "name": "",
  },
  {
    "id": 43,
    "langCode": "sl",
    "langDefault": false,
    "langOriginal": "Slovenščina",
    "langTranslated": "Slovenian",
    "name": "",
  },
  {
    "id": 10,
    "langCode": "es-419",
    "langDefault": false,
    "langOriginal": "Español (Latinoamérica)",
    "langTranslated": "Spanish (Latin America)",
    "name": "",
  },
  {
    "id": 11,
    "langCode": "es-ES",
    "langDefault": false,
    "langOriginal": "Español (España)",
    "langTranslated": "Spanish (Spain)",
    "name": "",
  },
  {
    "id": 46,
    "langCode": "sw",
    "langDefault": false,
    "langOriginal": "Kiswahili",
    "langTranslated": "Swahili",
    "name": "",
  },
  {
    "id": 45,
    "langCode": "sv",
    "langDefault": false,
    "langOriginal": "Svenska",
    "langTranslated": "Swedish",
    "name": "",
  },
  {
    "id": 47,
    "langCode": "ta",
    "langDefault": false,
    "langOriginal": "தமிழ்",
    "langTranslated": "T     amil",
    "name": "",
  },
  {
    "id": 48,
    "langCode": "te",
    "langDefault": false,
    "langOriginal": "తెలుగు",
    "langTranslated": "Telugu",
    "name": "",
  },
  {
    "id": 49,
    "langCode": "th",
    "langDefault": false,
    "langOriginal": "ไทย",
    "langTranslated": "Thai",
    "name": "",
  },
  {
    "id": 50,
    "langCode": "tr",
    "langDefault": false,
    "langOriginal": undefined,
    "langTranslated": "Turkish",
    "name": "",
  },
  {
    "id": 51,
    "langCode": "uk",
    "langDefault": false,
    "langOriginal": "Українська",
    "langTranslated": "Ukrainian",
    "name": "",
  },
  {
    "id": 52,
    "langCode": "vi",
    "langDefault": false,
    "langOriginal": "Tiếng Việt",
    "langTranslated": "Vietnamese",
    "name": "",
  },
]
`;
